pkg/tls/reloader_test.go

```diff
@@ -0,0 +1,953 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tls
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"math/big"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/apache/skywalking-banyandb/pkg/logger"
+)
+
+func generateSelfSignedCert(t *testing.T, commonName string) (certPEM, keyPEM []byte) {
+	t.Helper()
+
+	privateKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	require.NoError(t, err)
+
+	template := x509.Certificate{
+		SerialNumber: big.NewInt(1),
+		Subject: pkix.Name{
+			CommonName: commonName,
+		},
+		DNSNames:              []string{commonName},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().Add(time.Hour * 24),
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	certDER, err := x509.CreateCertificate(rand.Reader, &template, &template, &privateKey.PublicKey, privateKey)
+	require.NoError(t, err)
+
+	certPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certDER,
+	})
+
+	keyPEM = pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(privateKey),
+	})
+
+	return certPEM, keyPEM
+}
+
+func TestReloader_CertificateRotation(t *testing.T) {
+	// Create temporary directory for test files
+	tempDir, err := os.MkdirTemp("", "tls-test")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDir)
+
+	certFile := filepath.Join(tempDir, "cert.pem")
+	keyFile := filepath.Join(tempDir, "key.pem")
+
+	// Generate initial certificate
+	certPEM1, keyPEM1 := generateSelfSignedCert(t, "test1.local")
+	err = os.WriteFile(certFile, certPEM1, 0o600)
+	require.NoError(t, err)
+	err = os.WriteFile(keyFile, keyPEM1, 0o600)
+	require.NoError(t, err)
+
+	log := logger.GetLogger("tls-test")
+	reloader, err := NewReloader(certFile, keyFile, log)
+	require.NoError(t, err)
+	defer reloader.Stop()
+
+	// Start reloader
+	reloader.Start()
+
+	// Wait for initial certificate to be loaded
+	time.Sleep(100 * time.Millisecond)
+
+	// Verify initial certificate
+	tlsConfig := reloader.GetTLSConfig()
+	cert, err := tlsConfig.GetCertificate(nil)
+	require.NoError(t, err)
+	leafCert, err := x509.ParseCertificate(cert.Certificate[0])
+	require.NoError(t, err)
+	require.Equal(t, "test1.local", leafCert.Subject.CommonName)
+
+	// Create a new directory and move files there to trigger watcher
+	newDir := filepath.Join(tempDir, "new")
+	require.NoError(t, os.Mkdir(newDir, 0o755))
+
+	// Move files to new location
```

Comment:
Could you please explain why you're moving the files? It's not a user case we discussed before.